File: examples/scripts/python/add_external_data.py
#!/usr/bin/env python3
"""Model Health Python SDK — external data upload demo.

Demonstrates attaching external files (e.g. sensor data CSVs)
to an existing activity.

Usage:
    add_external_data.py [<api_key>]
"""

import os
import sys

from docopt import docopt

from modelhealth import (
    ExternalResultFile,
    ModelHealthError,
    ModelHealthClient,
)
from _prompts import pick_one
from _utils import load_api_key

# Activities created by the mobile app for internal use — exclude from lists.
_INTERNAL_ACTIVITY_NAMES = {"calibration", "neutral"}


def _connect(api_key):
    try:
        return ModelHealthClient(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")


def _pick_activity(client):
    print("\nFetching sessions...")
    try:
        sessions = client.session_list()
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch sessions: {exc}")

    if not sessions:
        sys.exit("No sessions found.")

    print(f"\n{len(sessions)} session(s):\n")
    session = pick_one(
        sessions,
        "Select session",
        lambda s: f"[{s.id}]  {s.session_name or '(unnamed)'}  —  subject: {s.name or '(unnamed)'}",
    )

    print(f"\nFetching activities for session {session.id}...")
    try:
        all_activities = client.activity_list(session)
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch activities: {exc}")

    activities = [a for a in all_activities if a.name not in _INTERNAL_ACTIVITY_NAMES]
    if not activities:
        sys.exit("No activities found in this session.")

    print(f"\n{len(activities)} activity/activities:\n")
    return pick_one(
        activities,
        "Select activity",
        lambda a: f"{a.name or a.id}  [{a.status}]",
    )


def _prompt_files():
    files = []
    print("\nEnter the files to attach (leave path blank to finish).")
    print("Tag:  a short identifier for the data source, e.g. 'my-force-plate'.")
    print()

    while True:
        path = input("  File path (or Enter to finish): ").strip()
        if not path:
            if not files:
                print("  At least one file is required.")
                continue
            break

        path = os.path.expanduser(path)
        if not os.path.isfile(path):
            print(f"  File not found: {path}")
            continue

        tag = input("  Tag for this file: ").strip()
        if not tag:
            print("  Tag must not be empty.")
            continue

        try:
            with open(path, "rb") as fh:
                data = fh.read()
        except OSError as exc:
            print(f"  Could not read file: {exc}")
            continue

        _, dot_ext = os.path.splitext(path)
        extension = dot_ext.lstrip(".") if dot_ext else input("  File extension (e.g. csv, bin): ").strip()
        if not extension:
            print("  Extension must not be empty.")
            continue

        files.append(ExternalResultFile(tag, extension, data))
        print(f"  Added: {os.path.basename(path)} (tag={tag!r}, extension={extension!r})")

    return files


if __name__ == "__main__":
    args = docopt(__doc__)
    api_key = load_api_key(args["<api_key>"])

    client = _connect(api_key)
    activity = _pick_activity(client)
    files = _prompt_files()

    print(f"\nUploading {len(files)} file(s)...")
    try:
        client.add_motion_data_to_activity(activity, files)
    except ModelHealthError as exc:
        sys.exit(f"Upload failed: {exc}")

    print("\nDone.")

File: examples/scripts/ts/add_external_data.ts
/**
 * Model Health TypeScript SDK — external data upload.
 * Mirrors examples/python/add_external_data.py.
 *
 * Usage:
 *   npx tsx add_external_data.ts [<api_key>]
 */

import * as fs from 'fs';
import * as path from 'path';
import { ModelHealthClient, type Activity, type ExternalResultFile } from '@modelhealth/modelhealth';
import {
  loadApiKey, INTERNAL_ACTIVITY_NAMES,
  pickOne, prompt, closePrompts,
} from './_shared.js';

async function main() {
  const args = process.argv.slice(2);
  const client = new ModelHealthClient({ apiKey: loadApiKey(args[0]), autoInit: false });
  await client.init();

  // Session
  console.log('\nFetching sessions...');
  const sessions = await client.sessionList();
  if (!sessions.length) { console.error('No sessions found.'); process.exit(1); }

  console.log(`\n${sessions.length} session(s):\n`);
  const session = await pickOne(sessions, 'Select session', s => {
    const sn = s.sessionName || '(unnamed)';
    const sub = s.name || '(unnamed)';
    return `[${s.id}]  ${sn}  —  subject: ${sub}`;
  });

  // Activities
  console.log(`\nFetching activities for session ${session.id}...`);
  const allActivities = await client.activityList(session.id);
  const activities = allActivities.filter(a => !INTERNAL_ACTIVITY_NAMES.has(a.name ?? ''));
  if (!activities.length) { console.error('No activities found in this session.'); process.exit(1); }

  console.log(`\n${activities.length} activity/activities:\n`);
  const activity = await pickOne(activities, 'Select activity', a => `${a.name ?? a.id}  [${a.status}]`);

  // Files
  const files = await promptFiles();

  console.log(`\nUploading ${files.length} file(s)...`);
  await client.addMotionDataToActivity(activity, files);
  console.log('\nDone.');
}

async function promptFiles(): Promise<ExternalResultFile[]> {
  const files: ExternalResultFile[] = [];
  console.log('\nEnter the files to attach (leave path blank to finish).');
  console.log("Tag:  a short identifier for the data source, e.g. 'my-force-plate'.");
  console.log();

  while (true) {
    const rawPath = (await prompt('  File path (or Enter to finish): ')).trim();
    if (!rawPath) {
      if (!files.length) { console.log('  At least one file is required.'); continue; }
      break;
    }

    const resolved = path.resolve(rawPath.replace(/^~/, process.env.HOME ?? '~'));
    if (!fs.existsSync(resolved)) { console.log(`  File not found: ${resolved}`); continue; }

    const tag = (await prompt('  Tag for this file: ')).trim();
    if (!tag) { console.log('  Tag must not be empty.'); continue; }

    let data: Buffer;
    try { data = fs.readFileSync(resolved); } catch (e) { console.log(`  Could not read file: ${e}`); continue; }

    const ext = path.extname(resolved).replace(/^\./, '');
    let extension = ext;
    if (!extension) {
      extension = (await prompt('  File extension (e.g. csv, bin): ')).trim();
      if (!extension) { console.log('  Extension must not be empty.'); continue; }
    }

    files.push({ tag, extension, data: new Uint8Array(data) });
    console.log(`  Added: ${path.basename(resolved)} (tag=${JSON.stringify(tag)}, extension=${JSON.stringify(extension)})`);
  }

  return files;
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);
